### DISCUSSION TITLE
using train.py

### DISCUSSION THREAD

**Discussion by jihk321** [ultralytics/yolov5#7570] [Q&A]
I am sorry for my poor English.
Custom data best.pt was created using yolov5s.pt.
I keep adding data, but as the amount of data increases, it takes a long time to learn and the mAP does not come out high.
When training, is it okay to use best.pt that I learned the weight value before?

**Comment by glenn-jocher** [comment]
@jihk321 👋 Hello! Thanks for asking about **resuming training**. YOLOv5 🚀 Learning Rate (LR) [schedulers](https://github.com/ultralytics/yolov5/blob/aa542ce6a65658ff931fee9bbab77c0145c152f0/train.py#L175-L181) follow predefined LR curves for the fixed number of `--epochs` defined at training start (`default=300`), and are designed to fall to a minimum LR on the final epoch for best training results. For this reason you can not modify the number of epochs once training has started.

<img width="850" alt="Screenshot 2022-04-10 at 11 11 51" src="https://user-images.githubusercontent.com/26833433/162611202-4e0b611f-6cca-4e43-94fc-bbc44e6447f8.png">

If your training was **interrupted** for any reason you may continue where you left off using the `--resume` argument. If your training fully completed, you can start a **new** training from any model using the `--weights` argument. Examples:

## Resume Single-GPU

You may not change settings when resuming, and **no additional arguments** other than `--resume` should be passed, with an optional path to the checkpoint you'd like to resume from. If no checkpoint is passed the most recently updated `last.pt` in your `yolov5/` directory is automatically found and used:
```bash
python train.py --resume  # automatically find latest checkpoint (searches yolov5/ directory)
python train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Resume Multi-GPU

[Multi-GPU DDP trainings](https://github.com/ultralytics/yolov5/issues/475) must be resumed with the same GPUs and DDP command, i.e. assuming 8 GPUs:
```bash
python -m torch.distributed.run --nproc_per_node 8 train.py --resume  # resume latest checkpoint
python -m torch.distributed.run --nproc_per_node 8 train.py --resume path/to/last.pt  # specify resume checkpoint
```

## Start from Pretrained

If you would like to **start training** from a fully trained model, use the `--weights` argument, not the `--resume` argument:
```bash
python train.py --weights path/to/best.pt  # start from pretrained model
```

Good luck 🍀 and let us know if you have any other questions!

**Reply by jihk321** [reply]
Thank you for your answer.
Is resume irrelevant if the dataset is updated or class is added?

**Reply by glenn-jocher** [reply]
@jihk321 if the class count changes you can start training from best.pt.

**Reply by jihk321** [reply]
> @jihk321 if the class count changes you can start training from best.pt.

Thanks!!!

**Comment by tleyden** [comment]
This is such a useful feature! 🤩  

I had a colab training run crash after 20 hours and was able to successfully resume from a saved checkpoint, it even kept track of the completed/total epoch numbers.

@glenn-jocher do you agree the CLI docs could use some improvement w.r.t resuming?  Currently it is:

```
    parser.add_argument('--resume', nargs='?', const=True, default=False, help='resume most recent training')
```

which misses some important details like:

1. It can take a path to a checkpoint
2. No other args need to be passed

Let me know if you'd take a PR.

**Reply by glenn-jocher** [reply]
@tleyden yes I agree, --resume could use better documentation 👍
